src/codegen/codegen_helper_visitor.cpp

```diff
@@ -276,19 +276,89 @@ void CodegenHelperVisitor::find_non_range_variables() {
+ * - Variable can not be range as well as state, it's redeclaration error
+ * - Variable can be parameter as well as range. Without range, parameter
+ *   is considered as global variable i.e. one value for all instances.
+ * - If a variable is only defined as RANGE and not in assgined or parameter
+ *   or state block then it's not printed.
+ * - Note that if variable property is different than the variable type. For example,
+ *   if variable has range property, it doesn't mean the variable is declared as RANGE.
+ *   Other variables like STATE and ASSIGNED block variables also get range property.
+ * - Also, there is difference between declaration order vs. definition order. For
+ *   example, POINTER variable in NEURON block is just declaration and doesn't
+ *   determine the order in which they will get printed. Below we query symbol table
+ *   and order all instance variables into certain order.
  */
 void CodegenHelperVisitor::find_range_variables() {
     /// comparator to decide the order based on definition
     auto comparator = [](const SymbolType& first, const SymbolType& second) -> bool {
         return first->get_definition_order() < second->get_definition_order();
     };
 
+    /// from symbols vector `vars`, remove all ion variables which are not ionic currents
+    auto remove_non_ioncur_vars = [](SymbolVectorType& vars, CodegenInfo& info) -> void {
+        vars.erase(std::remove_if(vars.begin(),
+                                  vars.end(),
+                                  [&](SymbolType& s) {
+                                      return info.is_ion_variable(s->get_name()) &&
+                                             !info.is_ionic_current(s->get_name());
+                                  }),
+                   vars.end());
+    };
+
+    /// if `seconday` vector contains any symbol that exist in the `primary` then remove it
+    auto remove_var_exist = [](SymbolVectorType& primary, SymbolVectorType& secondary) -> void {
+        secondary.erase(std::remove_if(secondary.begin(),
+                                       secondary.end(),
+                                       [&primary](const SymbolType& tosearch) {
+                                           return std::find_if(primary.begin(),
+                                                               primary.end(),
+                                                               // compare by symbol name
+                                                               [&tosearch](
+                                                                   const SymbolType& symbol) {
+                                                                   return tosearch->get_name() ==
+                                                                          symbol->get_name();
+                                                               }) != primary.end();
+                                       }),
+                        secondary.end());
+    };
+
```

Comment:
@tristan0x : is there any better pattern to achieve above? Or above is "fine"?

Reply:
I had a suggestion for better patterns :D